src/vm/i386/umthunkstub.S

```diff
@@ -173,25 +173,20 @@ LOCAL_LABEL(UMThunkStub_CopyStackArgs):
 LOCAL_LABEL(InitCopyStack):
     push    ecx
     push    edx
-    mov     eax, 0  // Offset in edi
-    mov     ebx, 0  // Offset in esi
-LOCAL_LABEL(CopyStack):
     mov     edx, dword ptr [ebp - UMThunkStub_UMENTRYTHUNK_OFFSET]
     mov     edx, dword ptr [edx + UMEntryThunk__m_pUMThunkMarshInfo]
-    mov     ecx, dword ptr [edx + UMThunkMarshInfo__m_ecxArgOffset]
-    cmp     ecx, ebx
+    mov     ebx, [edx + UMThunkMarshInfo__m_cbActualArgSize]
```

Comment:
Hmm, this looks like what I have suggested. But if it doesn't work, then I think there must be some other problem that just hides this problem and makes the test work.
Just look at a simple case - a function with a single parameter passed in ecx. I would expect that the m_ecxArgOffset would be set to zero, right? Then in your loop, you load m_cbActualArgSize to ebx. That would be 4, since that's the size of the single argument, right? And now you compare ebx to m_ecxArgOffset. It doesn't match, since the ebx is 4 and m_ecxArgOffset is 0. So the argument would be plaved to stack. At the end of the loop, we subtract 4 from ebx and since it becomes zero, exit the loop. So it ends up putting the argument to stack.
Can you please step through the code while running the FunctionPtrTest to see what's exactly going on?